GrandCentralBoard/Widgets/Bonus/Scene/BubbleScalingController.swift

```diff
@@ -0,0 +1,82 @@
+//
+//  BubbleScalingController.swift
+//  GrandCentralBoard
+//
+//  Created by Bartłomiej Chlebek on 07/04/16.
+//  Copyright © 2016 Oktawian Chojnacki. All rights reserved.
+//
+
+import SpriteKit
+
+protocol BubbleScalingControllerDelegate: class {
+    func bubbleScalingController(bubbleScalingController: BubbleScalingController, didScaleSpriteNodeUp spriteNode: SKSpriteNode)
+    func bubbleScalingController(bubbleScalingController: BubbleScalingController, didScaleSpriteNodeDown spriteNode: SKSpriteNode)
+}
+
+extension BubbleScalingControllerDelegate {
+    func bubbleScalingController(bubbleScalingController: BubbleScalingController, didScaleSpriteNodeUp spriteNode: SKSpriteNode) { }
+    func bubbleScalingController(bubbleScalingController: BubbleScalingController, didScaleSpriteNodeDown spriteNode: SKSpriteNode) { }
+}
+
+final class BubbleScalingController {
+    enum State {
+        case Idle
+        case ScalingUp
+        case ScaledUp
+        case ScalingDown
+    }
+    
+    private var state: State = .Idle
+    private weak var spriteNode: SKSpriteNode?
+    private var scaleDownTimer: NSTimer?
+    weak var delegate: BubbleScalingControllerDelegate?
+    
+    init(spriteNode: SKSpriteNode) {
+        self.spriteNode = spriteNode
+    }
+    
+    func scaleUp() {
+        switch self.state {
+        case .Idle: fallthrough
+        case .ScalingDown: self.performScaleUpAction()
+        case .ScaledUp: self.rescheduleScaleDownDeferTimer()
+        case .ScalingUp: return
+        }
+    }
+    
+    // MARK - Scale actions
+    
+    private func performScaleUpAction() {
+        guard let spriteNode = self.spriteNode else { return }
+        
+        let scaleUpAction = SKAction.scaleTo(2.3, duration: 0.5)
+        self.state = .ScalingUp
+        spriteNode.runAction(scaleUpAction, completion: {
+            self.state = .ScaledUp
+            self.rescheduleScaleDownDeferTimer()
+            self.delegate?.bubbleScalingController(self, didScaleSpriteNodeUp: spriteNode)
+        })
+    }
+    
+    @objc private func performScaleDownAction() {
+        guard let spriteNode = self.spriteNode else { return }
+        
+        let scaleDownAction = SKAction.scaleTo(1, duration: 0.5)
+        self.state = .ScalingDown
+        spriteNode.runAction(scaleDownAction, completion: {
+            self.state = .Idle
+            self.delegate?.bubbleScalingController(self, didScaleSpriteNodeDown: spriteNode)
+        })
+    }
+    
+    // MARK - Scale down timer
+    
+    private func rescheduleScaleDownDeferTimer() {
+        self.scaleDownTimer?.invalidate()
+        self.scaleDownTimer = NSTimer.scheduledTimerWithTimeInterval(15,
```

Comment:
This line create retains cycle in the project (always be careful when using `NSTimer`): 
1. You create NSTimer and keep strong or weak reference to it (it doesn't matter).
2. One you schedule it is added to the NSRunLoop which keeps strong reference to it.
3. Target of the NSTimer is retained by the Timer.

➡️ We have situation in which `NSRunLoop` keeps reference to the `NSTimer` which keeps reference to the bubble. The only way to break this cycle (and properly release `NSTimer` instance) it to call `invalidate` method on the timer. Personally, I would add invalidation of the timer to the implementation of the `performScaleDownAction` method.